src/utils/encoding/cid_de_cbor.rs

```diff
@@ -0,0 +1,185 @@
+// Copyright 2019-2023 ChainSafe Systems
+// SPDX-License-Identifier: Apache-2.0, MIT
+
+use crate::utils::encoding::from_slice_with_fallback;
+use cid::serde::BytesToCidVisitor;
+use cid::Cid;
+use core::fmt;
+use serde::de::{self, DeserializeSeed, SeqAccess, Visitor};
+use serde::Deserializer;
+
+/// Find and extract all the [`Cid`] from a `DAG_CBOR`-encoded blob without employing any
+/// intermediate recursive structures, eliminating unnecessary allocations.
+pub fn extract_cids(cbor_blob: &[u8]) -> anyhow::Result<Vec<Cid>> {
+    let cid_vec: CidVec = from_slice_with_fallback(cbor_blob)?;
+    Ok(cid_vec.into_inner())
+}
+
+/// [`CidVec`] allows for efficient zero-copy de-serialization of `DAG_CBOR`-encoded nodes into a
+/// vector of [`Cid`].
+#[derive(Default)]
+struct CidVec(Vec<Cid>);
+
+impl CidVec {
+    pub fn into_inner(self) -> Vec<Cid> {
+        self.0
+    }
+}
```

Comment:
```suggestion
/// Find and extract all the [`Cid`]s from a `DAG_CBOR`-encoded blob without employing any
/// intermediate allocations.
pub fn extract_cids(cbor_blob: &[u8]) -> anyhow::Result<Vec<Cid>> {
    let CidVec(v) = from_slice_with_fallback(cbor_blob)?;
    Ok(v)
}

struct CidVec(Vec<Cid>);
```

Reply:
This won't compile. I'll clean it up manually.